### PR TITLE
feat: Add option to startActivity on another display

### PR DESCRIPTION
**Pull request by chaooe**
https://github.com/appium/appium-espresso-driver/issues/546

### CODE REVIEW

**Comment by chaooe** [comment]
We can't touch on the another display as this parameter are only work on Android Q.
https://source.android.com/devices/tech/display/multi_display/input-routing

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt [hunk 1]
```diff
@@ -87,6 +88,8 @@ object ActivityHelpers {
             AndroidLogger.logger.info("Staring activity with custom options: $intentOptions")
             makeIntent(intentOptions)
         }
-        instrumentation.startActivitySync(intent)
+        val options = ActivityOptions.makeBasic()
+        if (displayId != null) options.launchDisplayId = displayId.toInt()
```
Comment (KazuCocoa) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: Only for API level 26+?
https://developer.android.com/reference/kotlin/android/app/ActivityOptions#setlaunchdisplayid

File: lib/desired-caps.js [hunk 2]
```diff
@@ -34,6 +34,9 @@ let espressoCapConstraints = {
   },
   intentOptions: {
     isObject: true
+  },
+  displayId: {
```
Comment (KazuCocoa) [inline lib/desired-caps.js]: How about `launchDisplayId` or `activityOptionLaunchDisplayId` etc to make the naming more clear?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt (hunk 1 above)
Comment (KazuCocoa) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: like https://github.com/appium/appium-espresso-driver/blob/d7b3f6cf27ec75741041c033d83728b6fcd01f96/espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/AutofillHelpers.kt#L40

Comment (KazuCocoa) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: You can also write like below:
```kotlin
displayId?.let {
    options.launchDisplayId = it.toInt()
}
```

**Comment by KazuCocoa** [comment]
You can add a test like https://github.com/appium/appium-espresso-driver/search?l=JavaScript&q=startActivity (In JS, but it simply calls the activity)

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt (hunk 1 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: Yes. It is 26+, let me add this

File: lib/desired-caps.js (hunk 2 above)
Comment (chaooe) [reply lib/desired-caps.js]: yes. launchDisplayId is more clear.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt [hunk 3]
```diff
@@ -41,7 +41,8 @@ class CreateSession : RequestHandler<SessionParams, Session> {
                 startActivity(
                         parsedCaps["appPackage"] as? String,
                         parsedCaps["appActivity"] as? String,
-                        parsedCaps["intentOptions"] as? Map<String, Any?>
+                        parsedCaps["intentOptions"] as? Map<String, Any?>,
+                        parsedCaps["displayId"] as? Number
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt]: I suppose it must be an integer, not just any number

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt [hunk 4]
```diff
@@ -41,7 +41,8 @@ class CreateSession : RequestHandler<SessionParams, Session> {
                 startActivity(
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt]: I would rather change startActivity signature to accept a data object. four arguments are too many

Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt]: Yes. I also want to have a data object. But i don't know how to write.

Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt]: You could simply reuse StartActivityParams one

**Comment by mykola-mokhnach** [comment]
I see that some integration tests started to fail. Could you please keep the old way of starting activity and only apply the new one when additional options are present?

**Comment by chaooe** [comment]
Yes. I will add it.

**Comment by chaooe** [comment]
When i am test the `npm run e2e-test`, i found an issue on MacOS

> dbug Espresso Beginning build with command './gradlew -PappiumTargetPackage=io.appium.android.apis assembleAndroidTest' in directory '/var/folders/s4/33jxs27n05z9d1g65ptqnn994hcf7q/T/espresso-server-10.13.2.111:5555'
dbug Espresso Output from Gradle will not be logged. To change this, use 'showGradleLog' desired capability
ERR! Espresso Unable to build Espresso server - Process ended with exitcode 1 (cmd: './gradlew -PappiumTargetPackage\=io.appium.android.apis assembleAndroidTest')
ERR! Espresso Gradle error message:
ERR! Espresso 
ERR! Espresso [STDERR] 错误: 找不到或无法加载主类 org.gradle.wrapper.GradleWrapperMain
dbug Espresso Deleting espresso session
dbug ADB Running '/Users/alex.zhao/Library/Android/sdk/platform-tools/adb -P 5037 -s 10.13.2.111\:5555 shell am force-stop io.appium.android.apis'
dbug Logcat Stopping logcat capture
dbug ADB Removing forwarded port socket connection: 8300 
dbug ADB Running '/Users/alex.zhao/Library/Android/sdk/platform-tools/adb -P 5037 -s 10.13.2.111\:5555 forward --remove tcp\:8300'
dbug MJSONWP Encountered internal error running command: A new session could not be created. Details: Unable to build Espresso server - Process ended with exitcode 1 (cmd: './gradlew -PappiumTargetPackage\=io.appium.android.apis assembleAndroidTest')
dbug MJSONWP Gradle error message:
dbug MJSONWP 
dbug MJSONWP [STDERR] 错误: 找不到或无法加载主类 org.gradle.wrapper.GradleWrapperMain. Check https://github.com/appium/appium-espresso-driver#troubleshooting regarding advanced session startup troubleshooting.
dbug MJSONWP Error: Unable to build Espresso server - Process ended with exitcode 1 (cmd: './gradlew -PappiumTargetPackage\=io.appium.android.apis assembleAndroidTest')
dbug MJSONWP Gradle error message:
dbug MJSONWP 
dbug MJSONWP [STDERR] 错误: 找不到或无法加载主类 org.gradle.wrapper.GradleWrapperMain. Check https://github.com/appium/appium-espresso-driver#troubleshooting regarding advanced session startup troubleshooting.
dbug MJSONWP     at Object.errorAndThrow (/Users/alex.zhao/Work/Auto/appium-alex/appium-espresso-driver/node_modules/_appium-support@2.41.0@appium-support/lib/logging.js:87:35)
dbug MJSONWP     at ServerBuilder.runBuildProcess (/Users/alex.zhao/Work/Auto/appium-alex/appium-espresso-driver/lib/server-builder.js:178:11)

My Android is 10.13.2.111:5555. when create a temp folder on /var/folder/T/xxx10.13.2.111:5555. It may caused by the folder name which have ":" in it.

**Comment by mykola-mokhnach** [comment]
@chaooe Oh, this is a fancy bug. I'll commit a fix soon

**Comment by chaooe** [comment]
are they the same?
`InstrumentationRegistry.getInstrumentation().startActivitySync(intent)`
and `InstrumentationRegistry.getInstrumentation().targetContext.startActivity(intent)`

I don't know why it failed again if they are the same.

**Comment by mykola-mokhnach** [comment]
`startActivitySync` does some additional stuff on top of `getTargetContext().startActivity`. 
Perhaps, the problem there is a race condition, since the emulator on CI VMs could sometimes be quite slow

```
public Activity startActivitySync(@NonNull Intent intent, @Nullable Bundle options) {
        validateNotAppThread();

        synchronized (mSync) {
            intent = new Intent(intent);

            ActivityInfo ai = intent.resolveActivityInfo(
                getTargetContext().getPackageManager(), 0);
            if (ai == null) {
                throw new RuntimeException("Unable to resolve activity for: " + intent);
            }
            String myProc = mThread.getProcessName();
            if (!ai.processName.equals(myProc)) {
                // todo: if this intent is ambiguous, look here to see if
                // there is a single match that is in our package.
                throw new RuntimeException("Intent in process "
                        + myProc + " resolved to different process "
                        + ai.processName + ": " + intent);
            }

            intent.setComponent(new ComponentName(
                    ai.applicationInfo.packageName, ai.name));
            final ActivityWaiter aw = new ActivityWaiter(intent);

            if (mWaitingActivities == null) {
                mWaitingActivities = new ArrayList();
            }
            mWaitingActivities.add(aw);

            getTargetContext().startActivity(intent, options);

            do {
                try {
                    mSync.wait();
                } catch (InterruptedException e) {
                }
            } while (mWaitingActivities.contains(aw));

            return aw.activity;
        }
    }
```

File: lib/desired-caps.js (hunk 2 above)
Comment (KazuCocoa) [reply lib/desired-caps.js]: Here also should be `launchDisplayId`

Comment (chaooe) [reply lib/desired-caps.js]: Yes. Fixed

**Comment by chaooe** [comment]
When i startActivitySync without options, that passed the Junit test. I have no idea about why it failed when i add options in it.
Pass log without options:
https://dev.azure.com/AppiumCI/Appium%20CI/_build/results?buildId=8571&view=logs&j=aee5d64f-51cb-51de-3064-efac91038b74
Failed log with options
https://dev.azure.com/AppiumCI/7a2a2919-d2b4-4720-a6c3-3a639f9bb2e0/_apis/build/builds/8563/logs/37
Do you have any suggestion?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/StartActivity.kt [hunk 5]
```diff
@@ -25,7 +25,7 @@ class StartActivity : RequestHandler<StartActivityParams, Void?> {
 
     @Throws(AppiumException::class)
     override fun handleInternal(params: StartActivityParams): Void? {
-        startActivity(params.appPackage, params.appActivity, params.optionalIntentArguments)
+        startActivity(params.appPackage, params.appActivity, params.optionalIntentArguments, params.launchDisplayId)
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/StartActivity.kt]: please change startActivity signature, so it accepts a single data object instead of multiple args

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt [hunk 6]
```diff
@@ -87,6 +89,19 @@ object ActivityHelpers {
             AndroidLogger.logger.info("Staring activity with custom options: $intentOptions")
             makeIntent(intentOptions)
         }
-        instrumentation.startActivitySync(intent)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: I would rewrite this code in the following manner to keep the tests working:

```kotlin

         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && displayId != null) {
             val options = ActivityOptions.makeBasic()
             options.launchDisplayId = displayId.toInt()
             startActivitySync(intent, options)
         } else {
             instrumentation.startActivitySync(intent)
         }      

```

The `startActivitySync` could be simply copied from SDK and adapted for Android below version P if needed. private class members may be extracted with reflection from the actual instrumentation instance (we have a convenience wrappers for these in `ReflectionUtils` and `KReflectionUtils`)

Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: My device is API 27 which have no Instrumention.StartActivitySync(intent, options), so i have to try this way.

Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: How about this?
image

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/StartActivity.kt (hunk 5 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/StartActivity.kt]: Something like startActivity(parms)? and add  startActivity in ActivityHelpers.kt to invoke origin startActivty?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt (hunk 6 above)
Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: I assume that not using startActivitySync is failing the tests. For now this solution would suffice though to not make it overcomplicated. I assume the displayId option is not going to be used much

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/StartActivity.kt (hunk 5 above)
Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/StartActivity.kt]: yes, something like that, the method should accept one param instead of four

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt (hunk 6 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: Yes. But it can help me to start my cases because of our primary display id is 1.

File: lib/desired-caps.js [hunk 7]
```diff
@@ -34,6 +34,9 @@ let espressoCapConstraints = {
   },
   intentOptions: {
     isObject: true
+  },
+  launchDisplayId: {
```
Comment (mykola-mokhnach) [inline lib/desired-caps.js]: what if we create an `activityOptions` object instead? My point is that ActivityOptions object in Android may contain several options, so there is no point in having a separate capability for each of these: https://android.googlesource.com/platform/frameworks/base/+/master/services/core/java/com/android/server/am/ActivityManagerShellCommand.java#487

Comment (chaooe) [reply lib/desired-caps.js]: Yes, That is what i wanted.. but that is out of my current skills. Can anyone help on this?

Comment (mykola-mokhnach) [reply lib/desired-caps.js]: that's easy. Simply follow the way it's done for `intentOptions`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt (hunk 6 above)
Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: yes, that is why I say it is fine, so you could be unblocked with your test

File: lib/desired-caps.js (hunk 7 above)
Comment (chaooe) [reply lib/desired-caps.js]: Let me try to create my first "hello world" by kt.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt [hunk 8]
```diff
@@ -38,11 +39,13 @@ class CreateSession : RequestHandler<SessionParams, Session> {
             val shouldLaunchApp = (parsedCaps["autoLaunch"] ?: true) as Boolean
             if (shouldLaunchApp) {
                 @Suppress("UNCHECKED_CAST")
-                startActivity(
+                val startParamss = StartActivityParams(
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt]: this intermediate variable is redundant

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt [hunk 9]
```diff
@@ -104,4 +104,10 @@ object ActivityHelpers {
             instrumentation.startActivitySync(intent)
         }
     }
+    fun startActivity(params: StartActivityParams) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: there is no need to create a new startActivity method. Simply change the signature for the existing one

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+ * limitations under the License.
+ */
+
+package io.appium.espressoserver.lib.helpers
+
+import android.app.ActivityOptions
+import kotlin.reflect.KClass
+import kotlin.reflect.full.cast
+
+private fun <T : Any> requireType(expectedType: KClass<T>, key: String, value: Any?): T {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: Please do not copypaste existing helpers. Simply move it to a common module, so they can be reused

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+ * @return The created intent
+ * @throws IllegalArgumentException if required options are missing or
+ * there is an issue with mapping value format
+ */
+
+fun makeActivityOptions(activityOptions: Map<String, Any?>): ActivityOptions {
+    val options = ActivityOptions.makeBasic()
+    val handlersMapping = mapOf<String, (key: String, value: Any?) -> Unit>(
+            "launchDisplayId" to fun(key, value) {
+                options.launchDisplayId = requireString(key, value).toInt()
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: this could be either string or int

File: lib/desired-caps.js [hunk 12]
```diff
@@ -35,8 +35,8 @@ let espressoCapConstraints = {
   intentOptions: {
     isObject: true
   },
-  launchDisplayId: {
-    isNumber: true
+  activityOptions: {
```
Comment (mykola-mokhnach) [inline lib/desired-caps.js]: 👍

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt [hunk 13]
```diff
@@ -87,6 +91,23 @@ object ActivityHelpers {
...
         }
-        instrumentation.startActivitySync(intent)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && activityOptions !=null) {
+            var options = makeActivityOptions(activityOptions)
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                instrumentation.startActivitySync(intent, options.toBundle())
+            } else {
+                instrumentation.targetContext.startActivity(intent, options.toBundle())
+            }
```
Comment (KazuCocoa) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: here can write like (can reduce var/val):

```kotlin
makeActivityOptions(activityOptions).let {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
        instrumentation.startActivitySync(intent, it.toBundle())
    } else {
        instrumentation.targetContext.startActivity(intent, it.toBundle())
    }
}

```

Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: Yes. When i am using this, it will have an error on junit test. Let me try again as you said

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt (hunk 10 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: yes. You are right, i found i don't need to copypaste again as they are in the same package

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt (hunk 8 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt]: How to write if i need to convert the data type to StartActivityParams? I don't know how to write

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt (hunk 9 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: How about this?
`

fun startActivity(params: StartActivityParams) {
        if (params.appActivity == null && params.optionalIntentArguments == null) {
            throw IllegalArgumentException("Either activity name or intent options must be set")
        }

        val instrumentation = InstrumentationRegistry.getInstrumentation()
        val intent = if (params.optionalIntentArguments == null) {
            val fullyQualifiedAppActivity = getFullyQualifiedActivityName(instrumentation, params.appPackage, params.appActivity!!)
            val defaultOptions = mapOf<String, Any>(
                    "action" to "ACTION_MAIN",
                    "flags" to "ACTIVITY_NEW_TASK",
                    "className" to fullyQualifiedAppActivity
            )
            AndroidLogger.logger.info("Starting activity '$fullyQualifiedAppActivity' " +
                    "with default options: $defaultOptions")
            makeIntent(defaultOptions)
        } else {
            AndroidLogger.logger.info("Staring activity with custom options: $intentOptions")
            makeIntent(params.optionalIntentArguments)
        }

        makeActivityOptions(params.optionalActivityArguments).let {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                instrumentation.startActivitySync(intent, it.toBundle())
            } else {
                instrumentation.targetContext.startActivity(intent, it.toBundle())
            }
        }
    }`

Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: yes, you got the idea

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt (hunk 8 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt]: How about this?
`
if (shouldLaunchApp) {
                @Suppress("UNCHECKED_CAST")
                startActivity(
                    StartActivityParams(
                            parsedCaps["appPackage"] as? String,
                            parsedCaps["appActivity"] as? String,
                            parsedCaps["intentOptions"] as? Map<String, Any?>,
                            parsedCaps["activityOptions"] as? Map<String, Any?>)
            )
            }
`

Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/CreateSession.kt]: fine with me

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt [hunk 14]
```diff
@@ -84,9 +88,18 @@ object ActivityHelpers {
                     "with default options: $defaultOptions")
             makeIntent(defaultOptions)
         } else {
-            AndroidLogger.logger.info("Staring activity with custom options: $intentOptions")
-            makeIntent(intentOptions)
+            AndroidLogger.logger.info("Staring activity with custom options: ${params.optionalIntentArguments}")
+            makeIntent(params.optionalIntentArguments)
+        }
+
+        makeActivityOptions(params.optionalActivityArguments).let {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: this won't work. There is no branch to execute if `params.optionalActivityArguments` is null

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+fun makeActivityOptions(activityOptions: Map<String, Any?>?): ActivityOptions {
+    val options = ActivityOptions.makeBasic()
+    val handlersMapping = mapOf<String, (key: String, value: Any?) -> Unit>(
+            "launchDisplayId" to fun(_, value) {
+                options.launchDisplayId = value.toString().toInt()
+            }
+    )
+    if (activityOptions !=null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: use `?.let` syntax

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+ * @return The created intent
+ * @throws IllegalArgumentException if required options are missing or
+ * there is an issue with mapping value format
+ */
+
+fun makeActivityOptions(activityOptions: Map<String, Any?>?): ActivityOptions {
+    val options = ActivityOptions.makeBasic()
+    val handlersMapping = mapOf<String, (key: String, value: Any?) -> Unit>(
+            "launchDisplayId" to fun(_, value) {
+                options.launchDisplayId = value.toString().toInt()
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: does this work for `Any?` type? I assume Kotlin might not even compile such expression. Also, I would expect a proper exceptions handling (e.g. We need to show users a hint about what types of value are expected)

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt (hunk 14 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: if `params.optionalActivityArguments` is null, it will return default options 
`val options = ActivityOptions.makeBasic()` from makeActivityOptions.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt (hunk 16 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: Yes. you are right. How about string for this value expected?

Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: make it string or number. both should be fine

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt (hunk 15 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: How about this?

> 

> activityOptions?.let {
        for ((optName: String, optValue: Any?) in it.entries) {
            val handler = handlersMapping[optName]
                    ?: throw IllegalArgumentException("The option named '$optName' is not known. " +
                            "Only the following options are supported: ${handlersMapping.keys}")
            handler(optName, optValue)
        }
    }

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt (hunk 16 above)
Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: Ok

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+ * @return The created intent
+ * @throws IllegalArgumentException if required options are missing or
+ * there is an issue with mapping value format
+ */
+
+fun makeActivityOptions(activityOptions: Map<String, String>?): ActivityOptions {
+    val options = ActivityOptions.makeBasic()
+    val handlersMapping = mapOf<String, (key: String, value: String) -> Unit>(
+            "launchDisplayId" to fun(_, value) {
+                options.launchDisplayId = value.toInt()
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: there is still no proper input handling here. I would expect the `launchDisplayId` value to be either string or int. Also, if toInt conversion fails then an unhandled number format exception is going to be thrown: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/to-int.html

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+            }
+    )
+    activityOptions?.let {
+        for ((optName: String, optValue: String) in it.entries) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: optValue must be `Any?`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/StartActivityParams.kt [hunk 19]
```diff
@@ -3,5 +3,6 @@ package io.appium.espressoserver.lib.model
 data class StartActivityParams(
     val appPackage: String? = null,
     val appActivity: String? = null,
-    val optionalIntentArguments: Map<String, Any?>? = null
+    val optionalIntentArguments: Map<String, Any?>? = null,
+    val optionalActivityArguments: Map<String, String>? = null
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/StartActivityParams.kt]: `Map<String, Any?>`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt [hunk 20]
```diff
@@ -84,9 +88,18 @@ object ActivityHelpers {
...
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                instrumentation.startActivitySync(intent, it.toBundle())
+            } else {
+                instrumentation.targetContext.startActivity(intent, it.toBundle())
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: I'm still not very convinced about direct `instrumentation.targetContext.startActivity` usage. The initial proposal was to only use it if `optionalActivityArguments` are set. If these are not set then the normal `instrumentation.startActivitySync(intent)` call should be used to preserve the backward compatibility

Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: Ok. Got it.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt [hunk 21]
```diff
@@ -92,11 +92,15 @@ object ActivityHelpers {
             makeIntent(params.optionalIntentArguments)
         }
 
-        makeActivityOptions(params.optionalActivityArguments).let {
-            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
-                instrumentation.startActivitySync(intent, it.toBundle())
-            } else {
-                instrumentation.targetContext.startActivity(intent, it.toBundle())
+        if (params.optionalActivityArguments ==null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelpers.kt]: Please format

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt [hunk 22]
```diff
@@ -28,15 +29,18 @@ import android.app.ActivityOptions
...
  */
 
-fun makeActivityOptions(activityOptions: Map<String, String>?): ActivityOptions {
+fun makeActivityOptions(activityOptions: Map<String, Any?>?): ActivityOptions {
     val options = ActivityOptions.makeBasic()
-    val handlersMapping = mapOf<String, (key: String, value: String) -> Unit>(
+    val handlersMapping = mapOf<String, (key: String, value: Any?) -> Unit>(
             "launchDisplayId" to fun(_, value) {
-                options.launchDisplayId = value.toInt()
+                require(String::class.isInstance(value)) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt]: we should be able to accept both string and int:

```
try {
  if (String::class.isInstance(value)) {
      options.launchDisplayId = String::class.cast(value).toInt()
      return
  } else if (Number::class.isInstance(value)) {
      options.launchDisplayId = Number::class.cast(value).toInt()
      return
  }
} catch (NumberFormatException: e) {
   // cannot convert to int
   throw IllegalArgumentException...
}
// not string or number passed
throw IllegalArgumentException...
```

Comment (chaooe) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt:36]: Thanks a lot, I will add and test it.

**Comment by chaooe** [comment]
How to fix the conflict. change the desired-caps.js as below?
`disableSuppressAccessibilityService: {
    isBoolean: true
  },
  activityOptions: {
    isObject: true
  }`

**Comment by mykola-mokhnach** [comment]
keep both options

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+
+import android.app.ActivityOptions
+import kotlin.reflect.full.cast
+
+/**
+ * Creates an options using the given activityOptions
+ *
+ * @param activityOptions mapping. The mapping can have the following keys:
+ * - `launchDisplayId`: display id which you want to launch, such as {"launchDisplayId":"1"}
+ * @return The created intent
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityOptionsHelpers.kt:27]: the returned value is not an intent

**Comment by mykola-mokhnach** [comment]
Unfortunately I cannot merge this PR unless the conflict is resolved

**Comment by chaooe** [comment]
fix conflict by this step https://stackoverflow.com/questions/38949951/how-to-solve-merge-conflicts-across-forks/38955036
